### PR TITLE
Rollup recent PRs

### PR DESCRIPTION
**Pull request by killercup**
This contains (in this order)

- Add .iter() method by impleting IntoIterator on reference of Range (#24)
- Direction trait to factorize Upwards and Downwards impls (#25)
- Use PhantomData (#22)

### CODE REVIEW

**Comment by killercup** [comment]
@kimsnj I've merged together the three open PRs. Can you have a look and tell me if I messed up some merges? There were quite a few conflicts.

File: src/lib.rs [hunk 1]
```diff
@@ -425,73 +445,43 @@ where
...
     current: Idx,
     limit: Idx,
-    #[allow(unused)]
-    direction: Direction,
+    direction: PhantomData<Direction>,
     init: bool,
 }
 
-/// Iterate over a reverse range
+/// Iterate over a the provided range
```
Comment (kimsnj) [inline src/lib.rs]: Merge type: "... over **a** the …",

**Review by kimsnj — COMMENTED** [review]
Except for the little typo, it seems all good :)

**Comment by killercup** [comment]
Thanks!

**Pull request merged**